aea/test_tools/test_protocol.py

```diff
@@ -0,0 +1,169 @@
+# -*- coding: utf-8 -*-
+# ------------------------------------------------------------------------------
+#
+#   Copyright 2022-2023 Valory AG
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+"""This module contains test case classes based on pytest for AEA protocol testing."""
+
+
+from abc import abstractmethod
+from typing import List, Type
+from unittest import TestCase
+from unittest.mock import patch
+
+import pytest
+
+from aea.common import Address
+from aea.mail.base import Envelope
+from aea.protocols.base import Message
+from aea.protocols.dialogue.base import Dialogue, Dialogues
+
+
+class BaseMessageTestCase(TestCase):
+    """Base class to test one message encode/decode"""
+
+    __test__ = False
+
+    MESSAGE_CLASS: Type[Message] = Message
+
+    def perform_message_test(self, msg: Message) -> None:  # nosec
+        """Test message encode/decode."""
+        msg.to = "receiver"
+        assert msg._is_consistent()  # pylint: disable=protected-access
+        envelope = Envelope(to=msg.to, sender="sender", message=msg)
+        envelope_bytes = envelope.encode()
+
+        actual_envelope = Envelope.decode(envelope_bytes)
+        expected_envelope = envelope
+
+        assert expected_envelope.to == actual_envelope.to
+        assert expected_envelope.sender == actual_envelope.sender
+        assert (
+            expected_envelope.protocol_specification_id
+            == actual_envelope.protocol_specification_id
+        )
+        assert expected_envelope.message != actual_envelope.message
+
+        actual_msg = self.MESSAGE_CLASS.serializer.decode(actual_envelope.message_bytes)
+        actual_msg.to = actual_envelope.to
+        actual_msg.sender = actual_envelope.sender
+        expected_msg = msg
+        assert expected_msg == actual_msg
```

Comment:
referencing this as I found it useful to define two methods
https://github.com/valory-xyz/open-aea/blob/d6c6c4d44284db7b7f7bd56bcf019bfb33999bf1/packages/valory/connections/test_libp2p/tests/base.py#L362-L373